aws-signer/src/main/scala/http4s/aws/middleware/AwsSigner.scala

```diff
@@ -0,0 +1,136 @@
+package http4s.aws.middleware
+
+import java.util.Date
+
+import cats.data.Kleisli
+import cats.effect.Sync
+import cats.implicits._
+import fs2.Stream
+import org.http4s.client.Client
+import org.http4s.{Header, Request}
+import java.nio.charset.StandardCharsets
+
+import BytesToHex.bytesToHex
+
+object AwsSigner {
+  def apply[F[_]: Sync](id: String, secret: String, zone: String, service: String)(
+      client: Client[F]): Client[F] = {
+    val signer = new AwsSigner(Key(id, secret), zone, service)
+    client.copy(open = Kleisli { req =>
+      signer[F](req, Sync[F].delay(new Date)).flatMap(client.open(_))
+    })
+  }
+
+  case class Key(id: String, secret: String)
+}
+
+/** Implements AWS request signing v4 as an http4s middleware.
+  * Signing described here: https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+  */
+class AwsSigner(key: AwsSigner.Key, zone: String, service: String) {
+  val Method = "AWS4-HMAC-SHA256"
+  val Charset = StandardCharsets.UTF_8
+
+  private def dateFormat(s: String) = {
+    val f = new java.text.SimpleDateFormat(s)
+    f.setTimeZone(java.util.TimeZone.getTimeZone("UTC"))
+    f
+  }
+
+  val fullDateFormat = dateFormat("YYYYMMdd'T'HHmmss'Z'")
+  val shortDateFormat = dateFormat("YYYYMMdd")
+
+  def hash(bytes: Array[Byte]) = {
+    val digest = java.security.MessageDigest.getInstance("SHA-256")
+    bytes.grouped(1024 * 16).foreach { chunk =>
+      digest.update(chunk)
+    }
+    digest.digest
+  }
+
+  def bytes(s: String) = s.getBytes(Charset)
+
+  def hmac(key: Array[Byte], data: Array[Byte]) = {
+    val algo = "HmacSHA256"
+    val hmac = javax.crypto.Mac.getInstance(algo)
+
+    hmac.init(new javax.crypto.spec.SecretKeySpec(key.toArray, algo))
+    hmac.doFinal(data.toArray)
+  }
+
+  def sign(string: String, date: java.util.Date) = {
+    val kSecret = bytes(s"AWS4${key.secret}")
+    val kDate = hmac(kSecret, bytes(shortDateFormat.format(date)))
+    val kRegion = hmac(kDate, bytes(zone))
+    val kService = hmac(kRegion, bytes(service))
+    val kSigning = hmac(kService, bytes("aws4_request"))
+    hmac(kSigning, bytes(string))
+  }
+
+  def apply[F[_]: Sync](request: Request[F], getDate: F[Date]): F[Request[F]] =
```

Comment:
I might have called this method `sign`.

Reply:
Continuing a conversation that GitHub has decided we're done with, where I proposed `Timer[F]` over `F[Date]`:

> I think that actually makes this harder to test, because you'd need to construct a whole timer, whereas this can be IO.pure(mydate)

> The intended "main" interface to this is via apply, where the user doesn't actually need to pick a date.

Oh, I see.  You're right, this isn't as consequential as I initially thought.

I guess I have a modest preference for `Timer` as a consistent way that time is parameterized elsewhere in cats-effect.  It's on its way to being slimmed down to `Clock`, which is two methods instead of one.  But two is still bigger than one, and I'd at least work in terms of `java.time.Instant` instead of the mutable and should-have-been-deprecated-a-decade-ago `java.util.Date`.

Reply:
will do